grpc/creds/creds.go

```diff
@@ -51,9 +63,73 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
+func (tc *transportCredentials) peerIsWhitelisted(peerState tls.ConnectionState) (bool, error) {
+	// If there's no whitelist, all clients are OK
+	if tc.whitelist == nil {
+		return true, nil
+	}
+
+	// Otherwise its time to start inspecting the peer's `VerifiedChains`
+	chains := peerState.VerifiedChains
+	if len(chains) < 1 {
+		return false, fmt.Errorf("boulder/grpc/creds: peer had zero VerifiedChains")
+	}
+
+	/*
+	 * For each of the peer's verified chains we can look at the chain's leaf
+	 * certificate and check whether the subject common name is in the whitelist.
+	 * At least one chain must have a leaf certificate with a subject CN that
+	 * matches the whitelist
+	 *
+	 * Its important we process `VerifiedChains` instead of processing
+	 * `PeerCertificates` to ensure that we match the subject CN of the
+	 * leaf certificate that was verified in `conn.Handshake()`. To do otherwise
+	 * would allow an attacker to include a whitelisted certificate in
+	 * `PeerCertificates` that matched the whitelist but wasn't used in the chain
+	 * the server validated.
+	 */
+	var whitelisted bool
+	for _, chain := range chains {
+		leafSubjectCN := chain[0].Subject.CommonName
+		if _, ok := tc.whitelist[leafSubjectCN]; ok {
+			whitelisted = true
+		}
+	}
+
+	// If none of the peer's validated chains had a leaf certificate with a
+	// whitelisted CN then we have to reject the connection
+	if !whitelisted {
+		return false, fmt.Errorf(
```

Comment:
One quick thing, this function seems like it could just as easily return just a `error` instead of `bool, error`. The check below would then just be

```
if err := tc.peerIsWhitelisted(conn.ConnectionState()); err != nil {
  ...
}
```